cpp/include/raft/random/rmat_rectangular_generator.cuh

```diff
@@ -18,51 +18,366 @@
 
 #include "detail/rmat_rectangular_generator.cuh"
 
+#include <optional>
+#include <raft/core/device_mdspan.hpp>
+#include <raft/core/handle.hpp>
+#include <variant>
+
 namespace raft::random {
 
 /**
- * @brief Generate RMAT for a rectangular shaped adjacency matrices (useful when
+ * @brief Type of the output vector(s) parameter for `rmat_rectangular_gen`
+ *   (see below).
+ *
+ * @tparam IdxT Type of each node index; must be integral.
+ *
+ * Users can provide either `out` by itself, (`out_src` and `out_dst`)
+ * together, or all three (`out`, `out_src`, and `out_dst`).
+ * This class prevents users from doing anything other than that.
+ * It also checks compatibility of dimensions at run time.
+ *
+ * The following examples show how to create an output parameter.
+ *
+ * @code
+ * rmat_rectangular_gen_output<size_t> output1(out);
+ * rmat_rectangular_gen_output<size_t> output2(out_src, out_dst);
+ * rmat_rectangular_gen_output<size_t> output3(out, out_src, out_dst);
+ * @endcode
+ *
+ * @{
+ */
+template <typename IdxT>
```

Comment:
@cjnolet I'll be happy to simplify this.  The only reason for the complicated design was that the current interface appears to let users provide all three arguments (`out`, `src`, and `dst`).  If that's not meaningful or a use case that we want to support, I'll be happy to get rid of this and just use overloads to express the two cases.

Reply:
To clarify: Do users actually want to be able to write to both output formats (`out`, and (`src` and `dst`)) with a single function call?  If not, then just providing two overloads (one taking `out`, and the other taking `src` and `dst`) would make the most sense.

Reply:
@mhoemmen 

My intuition is telling me that the additional outputs were meant to be used as either the 2-column format or the separate arrays together, but that's just my intuition based on the redundancy of the data. 

@teju85, do you see any reason a user would want to generate, say, out+src, out+dst, or out+src+dst or is it more likely that a user would want either out or src+dst?

Reply:
This design allows users to automatically generate both AoS and SoA format of the edges. So, I'd expect users to pass either `out` or `dst, src` or both.

Reply:
@teju85 Thanks for explaining! 

Do we actually want to support the use case where users pass in all three arrays?

Reply:
Good question! I wasn't sure whether we really need to support that case. So, I just assumed that the users might need to pass all three of them and thus also supported that case [here](https://github.com/rapidsai/raft/blob/branch-22.10/cpp/include/raft/random/detail/rmat_rectangular_generator.cuh#L57).

Reply:
Is that case causing issues with wrapping this API with mdspan?

Reply:
@teju85 That case (all three arrays nonnull) was the only case being tested, before mdspan.  Introducing mdspan lets the function check the lengths of the arguments.  Before, the function just assumed that nonnull pointers had the right number of entries (`n_edges` for `src` and `dst`, or 2 x `n_edges` for `out`).

Reply:
In that sense, it's actually easier to support that use case (all three arrays provided) with mdspan, because we can check that all the array lengths are consistent.  This PR does that (checks the lengths).  It sounds like the only question is whether to use overloads or this extra class as the way to let users specify whatever arrays they want to specify.  It's trivial to go from the current design to overloads (I would just keep the extra class as an implementation detail and let it do the run-time checks).

Reply:
> It sounds like the only question is whether to use overloads or this extra class as the way to let users specify whatever arrays they want to specify. It's not hard to go from the current design to overloads (I would just keep the extra class as an implementation detail and let it do the run-time checks).

I'm not against using the classes to consolidate the outputs. I was proposing the overloads as a way to simplify the outputs in the even that we are not expecting users to provide all 3 outputs at once but if that's a use-case we want to support then I'm fine w/ the classes. We should still move them to their own `*_types.hpp` file, though.

Reply:
@cjnolet I could just express the same idea through overloads:

* `out` only,
* `src` and `dst` only, and
* `out`, `src`, and `dst`.

There's no extra safety benefit to exposing the extra class to users.  It's just a way to consolidate error checking.  The overloads wouldn't be ambiguous, because we wouldn't need `std::optional` for any of them, and because each case has a different number of parameters.

Reply:
That works too.

Reply:
Thanks all for the feedback!  I've made `rmat_rectangular_gen_output` an implementation detail in the `detail` namespace, and added overloads for each of the output cases (1, 2, or 3 output vectors).